### PR TITLE
updated cli tool to be used to generate openapi document

### LINKED ISSUE #303: CLI not working due to an unconfigured mock
**Describe the issue**

```
Object reference not set to an instance of an object.
   at Microsoft.Azure.WebJobs.Extensions.OpenApi.Document.Build(Assembly assembly, OpenApiVersionType version) in azure-functions-openapi-extension/src/Microsoft.Azure.WebJobs.Extensions.OpenApi/Document.cs:line 137
   at Microsoft.Azure.WebJobs.Extensions.OpenApi.Document.Build(String assemblyPath, OpenApiVersionType version) in azure-functions-openapi-extension/src/Microsoft.Azure.WebJobs.Extensions.OpenApi/Document.cs:line 124
   at Microsoft.Azure.WebJobs.Extensions.OpenApi.CLI.Program.Generate(String project, String configuration, String target, OpenApiVersionType version, OpenApiFormatType format, String output, Boolean console) in azure-functions-openapi-extension/src/Microsoft.Azure.WebJobs.Extensions.OpenApi.CLI/Program.cs:line 90
```

A Mock is constructed for [`req`](https://github.com/Azure/azure-functions-openapi-extension/blob/main/src/Microsoft.Azure.WebJobs.Extensions.OpenApi.CLI/Program.cs#L68) without providing a getter for `req.Query`. It fails eventually when it executes [`this._req.Query["tags"]`](https://github.com/Azure/azure-functions-openapi-extension/blob/4ab63c222d914ff131577ebd9d080ff5c272069a/src/Microsoft.Azure.WebJobs.Extensions.OpenApi/Document.cs#L137).

**To Reproduce**
1. Have a working project (i.e. the web interface works)
2. Run the CLI (e.g. `-p ../My.Project -v v3 -f yaml -t netcoreapp3.1 --console`
3. Observe it fails

**Expected behavior**
It generates a Swagger specification.

**Screenshots**
N/a

**Environment (please complete the following information, if applicable):**
- OS: macOS 12.0.1
- .NET SDK 5.0.402
- 0.9.0-preview (from source)

**Additional context**
N/a

### PR DESCRIPTION
**Pull request by Nibell**
Changes proposed in this pull request:

-  Updated to working cli to generate openapi document and handle multiple versions

Have not added build to generate console app, but added support for win, linux and mac os
Resolves #303, #305
Ready for review @justinyoo

### CODE REVIEW

**Comment by Nibell** [comment]
Have not added build to generate console app, but added support for win, linux and mac os

**Comment by Nibell** [comment]
Resolves #303, #305

**Comment by justinyoo** [comment]
@Nibell Thanks for the PR! I'll take a look and get back to you.

**Comment by Fazer01** [comment]
Hi @Nibell! Nice one on the refactor to integrate working CLI tooling to generate swagger file. 
Currently we use Azure Functions v3 and are searching for a good way to generate the swagger file after building the function-app to incorporate automatic updates to the APIM we also use. 

Now we wait for @justinyoo to get back to you.

**Comment by Nibell** [comment]
Updated PR to :NET 6 @justinyoo

File: src/Microsoft.Azure.WebJobs.Extensions.OpenApi.CLI/ConsoleApp/GenerateSwaggerConsoleApp.cs [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+            [FromService] ICustomOpenApiCreator customOpenApiCreator,
+            [FromService] ICustomOpenApiWriter customOpenApiWriter,
+            [Option('p', Description = "Api Project path. Default is current directory")]
+            string project = ".",
+            [Option('a', Description = "ApiBaseUrl. Default is 'localhost'")]
+            string apibaseurl = "localhost",
+            [Option('c', Description = "Configuration. Default is 'Debug'")]
+            string configuration = "Debug",
+            [Option('t', Description = "Target framework. Default is 'net5.0'")]
+            string target = "net5.0",
```
Comment (justinyoo) [inline src/Microsoft.Azure.WebJobs.Extensions.OpenApi.CLI/ConsoleApp/GenerateSwaggerConsoleApp.cs:32]: Let's set the default value of `net6.0` as it's LTS.

File: src/Microsoft.Azure.WebJobs.Extensions.OpenApi.CLI/ConsoleApp/GenerateSwaggerConsoleApp.cs [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+            [Option('v', Description = "OpenAPI spec version. Value can be either 'v2' or 'v3'. Default is 'v3'")]
+            OpenApiVersionType version = OpenApiVersionType.V3,
```
Comment (justinyoo) [inline src/Microsoft.Azure.WebJobs.Extensions.OpenApi.CLI/ConsoleApp/GenerateSwaggerConsoleApp.cs:34]: Let's set the default value of 'V2' as the existing library uses `V2` as default. Once the direction changes, then we can change it to `V3`.

File: src/Microsoft.Azure.WebJobs.Extensions.OpenApi.CLI/Extension/ProjectPathExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,73 @@
+using System;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using System.Xml;
+
+namespace Microsoft.Azure.WebJobs.Extensions.OpenApi.CLI.Extension
```
Comment (justinyoo) [inline src/Microsoft.Azure.WebJobs.Extensions.OpenApi.CLI/Extension/ProjectPathExtensions.cs:7]: Can you change the directory name and namespace from `Extension` to `Extensions` by following the conventions?

File: src/Microsoft.Azure.WebJobs.Extensions.OpenApi.CLI/Extension/SetupHostExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+using System.Linq;
+using System.Reflection;
+using Microsoft.Azure.WebJobs.Extensions.OpenApi.Configurations.AppSettings.Extensions;
+using Microsoft.Azure.WebJobs.Extensions.OpenApi.Core.Abstractions;
+using Microsoft.Azure.WebJobs.Extensions.OpenApi.Core.Configurations;
+using Microsoft.Azure.WebJobs.Extensions.OpenApi.Core.Extensions;
+using Microsoft.Extensions.Configuration;
+using Microsoft.OpenApi.Models;
+
+namespace Microsoft.Azure.WebJobs.Extensions.OpenApi.CLI.Extension
```
Comment (justinyoo) [inline src/Microsoft.Azure.WebJobs.Extensions.OpenApi.CLI/Extension/SetupHostExtensions.cs:11]: Can you change the directory name and namespace from `Extension` to `Extensions` by following the conventions?

File: src/Microsoft.Azure.WebJobs.Extensions.OpenApi.CLI/Model/ApiMock.cs [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+
+namespace Microsoft.Azure.WebJobs.Extensions.OpenApi.CLI.Model
+{
+    public class ApiMock
+    {
+        public string CompiledPath { set; get; }
+        public string CompiledDllPath { set; get; }
+        public HttpSettings HttpSettings { set; get; }
+        public OpenApiInfo OpenApiInfo { set; get; }
+    }
```
Comment (justinyoo) [inline src/Microsoft.Azure.WebJobs.Extensions.OpenApi.CLI/Model/ApiMock.cs:12]: Can we put XML comments on the class and each property?

File: src/Microsoft.Azure.WebJobs.Extensions.OpenApi.CLI/Service/Impl/CustomApiMockCreator.cs [hunk 6]
```diff
@@ -0,0 +1,41 @@
+using System;
+using Microsoft.Azure.WebJobs.Extensions.OpenApi.CLI.Extension;
+using Microsoft.Azure.WebJobs.Extensions.OpenApi.CLI.Model;
+using Microsoft.Azure.WebJobs.Extensions.OpenApi.CLI.Service.Interface;
+
+namespace Microsoft.Azure.WebJobs.Extensions.OpenApi.CLI.Service.Impl
```
Comment (justinyoo) [inline src/Microsoft.Azure.WebJobs.Extensions.OpenApi.CLI/Service/Impl/CustomApiMockCreator.cs:6]: Can we change the folder structure from `Service.Impl` to `Services` and also change the namespace accordingly?

File: src/Microsoft.Azure.WebJobs.Extensions.OpenApi.CLI/Service/Impl/CustomOpenApiCreator.cs [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+using Microsoft.Azure.WebJobs.Extensions.OpenApi.Core.Abstractions;
+using Microsoft.Azure.WebJobs.Extensions.OpenApi.Core.Configurations;
+using Microsoft.Azure.WebJobs.Extensions.OpenApi.Core.Enums;
+using Microsoft.Azure.WebJobs.Extensions.OpenApi.Core.Extensions;
+using Microsoft.Azure.WebJobs.Extensions.OpenApi.Core.Visitors;
+using Microsoft.OpenApi.Models;
+using Moq;
+using Newtonsoft.Json.Serialization;
+
+namespace Microsoft.Azure.WebJobs.Extensions.OpenApi.CLI.Service.Impl
```
Comment (justinyoo) [inline src/Microsoft.Azure.WebJobs.Extensions.OpenApi.CLI/Service/Impl/CustomOpenApiCreator.cs:15]: Can we change the folder structure from `Service.Impl` to `Services` and also change the namespace accordingly?

File: src/Microsoft.Azure.WebJobs.Extensions.OpenApi.CLI/Service/Impl/CustomOpenApiWriter.cs [hunk 8]
```diff
@@ -0,0 +1,21 @@
+using System.IO;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.Azure.WebJobs.Extensions.OpenApi.CLI.Extension;
+using Microsoft.Azure.WebJobs.Extensions.OpenApi.CLI.Service.Interface;
+using Microsoft.Azure.WebJobs.Extensions.OpenApi.Core.Enums;
+using Microsoft.Azure.WebJobs.Extensions.OpenApi.Core.Extensions;
+
+namespace Microsoft.Azure.WebJobs.Extensions.OpenApi.CLI.Service.Impl
```
Comment (justinyoo) [inline src/Microsoft.Azure.WebJobs.Extensions.OpenApi.CLI/Service/Impl/CustomOpenApiWriter.cs:9]: Can we change the folder structure from `Service.Impl` to `Services` and also change the namespace accordingly?

File: src/Microsoft.Azure.WebJobs.Extensions.OpenApi.CLI/Service/Interface/ICustomApiMockCreator.cs [hunk 9]
```diff
@@ -0,0 +1,9 @@
+using Microsoft.Azure.WebJobs.Extensions.OpenApi.CLI.Model;
+
+namespace Microsoft.Azure.WebJobs.Extensions.OpenApi.CLI.Service.Interface
```
Comment (justinyoo) [inline src/Microsoft.Azure.WebJobs.Extensions.OpenApi.CLI/Service/Interface/ICustomApiMockCreator.cs:3]: Can we change the folder structure from `Service.Interface` to `Abstractions` and also change the namespace accordingly?

File: src/Microsoft.Azure.WebJobs.Extensions.OpenApi.CLI/Service/Interface/ICustomOpenApiCreator.cs [hunk 10]
```diff
@@ -0,0 +1,17 @@
+using System.Threading.Tasks;
+using Microsoft.Azure.WebJobs.Extensions.OpenApi.Core.Enums;
+using Microsoft.OpenApi.Models;
+
+namespace Microsoft.Azure.WebJobs.Extensions.OpenApi.CLI.Service.Interface
```
Comment (justinyoo) [inline src/Microsoft.Azure.WebJobs.Extensions.OpenApi.CLI/Service/Interface/ICustomOpenApiCreator.cs:5]: Can we change the folder structure from `Service.Interface` to `Abstractions` and also change the namespace accordingly?

File: src/Microsoft.Azure.WebJobs.Extensions.OpenApi.CLI/Service/Interface/ICustomOpenApiWriter.cs [hunk 11]
```diff
@@ -0,0 +1,10 @@
+using System.Threading.Tasks;
+using Microsoft.Azure.WebJobs.Extensions.OpenApi.Core.Enums;
+
+namespace Microsoft.Azure.WebJobs.Extensions.OpenApi.CLI.Service.Interface
```
Comment (justinyoo) [inline src/Microsoft.Azure.WebJobs.Extensions.OpenApi.CLI/Service/Interface/ICustomOpenApiWriter.cs:4]: Can we change the folder structure from `Service.Interface` to `Abstractions` and also change the namespace accordingly?

**Review by justinyoo — CHANGES_REQUESTED** [review]
@Nibell Thanks for the long wait! It looks awesome! I left several comments for structuring.

As it's completely re-written, I need to take further look. Can you also make sure that it doesn't throw an error but shows the help message when it is run by itself, meaning `azfuncopenapi.exe` with no arguments?

File: src/Microsoft.Azure.WebJobs.Extensions.OpenApi.CLI/Service/Impl/CustomOpenApiCreator.cs (hunk 7 above)
Comment (Nibell) [reply src/Microsoft.Azure.WebJobs.Extensions.OpenApi.CLI/Service/Impl/CustomOpenApiCreator.cs]: Changed to match request in new commit

File: src/Microsoft.Azure.WebJobs.Extensions.OpenApi.CLI/Service/Impl/CustomOpenApiWriter.cs (hunk 8 above)
Comment (Nibell) [reply src/Microsoft.Azure.WebJobs.Extensions.OpenApi.CLI/Service/Impl/CustomOpenApiWriter.cs]: Changed to match request in new commit

File: src/Microsoft.Azure.WebJobs.Extensions.OpenApi.CLI/Service/Interface/ICustomApiMockCreator.cs (hunk 9 above)
Comment (Nibell) [reply src/Microsoft.Azure.WebJobs.Extensions.OpenApi.CLI/Service/Interface/ICustomApiMockCreator.cs]: Changed to match request in new commit

File: src/Microsoft.Azure.WebJobs.Extensions.OpenApi.CLI/Service/Interface/ICustomOpenApiWriter.cs (hunk 11 above)
Comment (Nibell) [reply src/Microsoft.Azure.WebJobs.Extensions.OpenApi.CLI/Service/Interface/ICustomOpenApiWriter.cs]: Changed to match request in new commit

File: src/Microsoft.Azure.WebJobs.Extensions.OpenApi.CLI/Service/Interface/ICustomOpenApiCreator.cs (hunk 10 above)
Comment (Nibell) [reply src/Microsoft.Azure.WebJobs.Extensions.OpenApi.CLI/Service/Interface/ICustomOpenApiCreator.cs]: Changed to match request in new commit

File: src/Microsoft.Azure.WebJobs.Extensions.OpenApi.CLI/Extension/SetupHostExtensions.cs (hunk 4 above)
Comment (Nibell) [reply src/Microsoft.Azure.WebJobs.Extensions.OpenApi.CLI/Extension/SetupHostExtensions.cs]: Changed to match request in new commit

File: src/Microsoft.Azure.WebJobs.Extensions.OpenApi.CLI/Extension/ProjectPathExtensions.cs (hunk 3 above)
Comment (Nibell) [reply src/Microsoft.Azure.WebJobs.Extensions.OpenApi.CLI/Extension/ProjectPathExtensions.cs]: Changed to match request in new commit

File: src/Microsoft.Azure.WebJobs.Extensions.OpenApi.CLI/ConsoleApp/GenerateSwaggerConsoleApp.cs (hunk 1 above)
Comment (Nibell) [reply src/Microsoft.Azure.WebJobs.Extensions.OpenApi.CLI/ConsoleApp/GenerateSwaggerConsoleApp.cs]: Changed to net6.0 in new commit

File: src/Microsoft.Azure.WebJobs.Extensions.OpenApi.CLI/ConsoleApp/GenerateSwaggerConsoleApp.cs (hunk 2 above)
Comment (Nibell) [reply src/Microsoft.Azure.WebJobs.Extensions.OpenApi.CLI/ConsoleApp/GenerateSwaggerConsoleApp.cs]: Changed to v2 in new commit

File: src/Microsoft.Azure.WebJobs.Extensions.OpenApi.CLI/Service/Impl/CustomApiMockCreator.cs (hunk 6 above)
Comment (Nibell) [reply src/Microsoft.Azure.WebJobs.Extensions.OpenApi.CLI/Service/Impl/CustomApiMockCreator.cs]: Changed to match request in new commit

**Review by justinyoo — CHANGES_REQUESTED** [review]
@Nibell I'm still experiencing the exception when I run this app with **NO** argument.

```
azfuncopenapi.exe
```

Could you double check for this?

**Comment by justinyoo** [comment]
@Nibell Here's what I ran:

```
azfuncopenapi.exe -p "C:\path\to\functionapp" -v v3
```

But I get the exception:

```
Could not load file or assembly 'Microsoft.Azure.WebJobs, Version=3.0.23.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35'. The system cannot find the file specified.
```

Can you reproduce this error on your end? The function app path is completely outside this repo.

**Comment by ThomasBleijendaal** [comment]
@justinyoo I can reproduce the error (Microsoft.Azure.WebJobs version 3.0.31 is missing).

There is a stray package reference in the CLI's csproj (`<PackageReference Include="Microsoft.Azure.WebJobs" Version="3.0.5" />`). If I update that to 3.0.31 it fixes that error, but I get 
```
Could not load file or assembly 'Microsoft.Azure.WebJobs.Extensions.Http, Version=3.0.2.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35'. The system cannot find the file specified.
```
after that. It looks like some dependencies are not loaded into memory. I don't think it is a good idea to add those missing items as package as it tightly couples the CLI tool with the exact versions used by the function app of which the open api docs are generated.

**Comment by coolhome** [comment]
@ThomasBleijendaal 
To fix the out-of-process/in-process coupling would it make sense to make `Microsoft.Azure.Functions.Worker.Extensions.OpenApi.Cli` for Out-of-process?

**Comment by justinyoo** [comment]
As the PR, #489, has been merged, you can generate the OpenAPI doc on-the-fly, within the GitHub Actions workflow.

**Comment by JeroenvdBurg** [comment]
I think it is really a pitty that this PR has been closed. We would definitly use it, the on the fly generation really feels cumbersome. 

@justinyoo could you consider re-opening this and make it only support out-of-proc functions? As with the new azure function roadmap the webjobs functions are not supported  in the future anyway.

**Pull request closed without merging**